core/data/src/main/kotlin/com/google/samples/apps/nowinandroid/core/data/util/TimeZoneMonitor.kt

```diff
@@ -0,0 +1,94 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.samples.apps.nowinandroid.core.data.util
+
+import android.content.BroadcastReceiver
+import android.content.Context
+import android.content.Intent
+import android.content.IntentFilter
+import android.os.Build.VERSION
+import android.os.Build.VERSION_CODES
+import androidx.core.os.trace
+import com.google.samples.apps.nowinandroid.core.network.Dispatcher
+import com.google.samples.apps.nowinandroid.core.network.NiaDispatchers.IO
+import com.google.samples.apps.nowinandroid.core.network.di.ApplicationScope
+import dagger.hilt.android.qualifiers.ApplicationContext
+import kotlinx.coroutines.CoroutineDispatcher
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.channels.awaitClose
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.SharedFlow
+import kotlinx.coroutines.flow.SharingStarted
+import kotlinx.coroutines.flow.callbackFlow
+import kotlinx.coroutines.flow.flowOn
+import kotlinx.coroutines.flow.shareIn
+import java.time.ZoneId
+import javax.inject.Inject
+import javax.inject.Singleton
+
+/**
+ * Utility for reporting current timezone the device has set.
+ * It always emits at least once with default setting and then for each TZ change.
+ */
+interface TimeZoneMonitor {
+    val currentZoneId: Flow<ZoneId>
+}
+
+@Singleton
+internal class TimeZoneBroadcastMonitor @Inject constructor(
+    @ApplicationContext private val context: Context,
+    @ApplicationScope appScope: CoroutineScope,
+    @Dispatcher(IO) private val ioDispatcher: CoroutineDispatcher,
+) : TimeZoneMonitor {
+
+    override val currentZoneId: SharedFlow<ZoneId> =
+        callbackFlow<ZoneId> {
+            // Send the default time zone first.
+            trySend(ZoneId.systemDefault())
+
+            // Registers BroadcastReceiver for the TimeZone changes
+            val receiver = object : BroadcastReceiver() {
+                override fun onReceive(context: Context, intent: Intent) {
+                    if (intent.action != Intent.ACTION_TIMEZONE_CHANGED) return
+
+                    val zoneIdFromIntent = if (VERSION.SDK_INT < VERSION_CODES.R) {
+                        null
+                    } else {
+                        // Starting Android R we also get the new TimeZone.
+                        intent.getStringExtra(Intent.EXTRA_TIMEZONE)?.let { zoneId ->
+                            // We need to convert it from java.util.Timezone to java.time.ZoneId
+                            ZoneId.of(zoneId, ZoneId.SHORT_IDS)
+                        }
+                    }
+
+                    // If there isn't a zoneId in the intent, fallback to the systemDefault, which should also reflect the change
+                    trySend(zoneIdFromIntent ?: ZoneId.systemDefault())
+                }
+            }
+
+            trace("TimeZoneBroadcastReceiver.register") {
```

Comment:
In Android Studio, I'm seeing this `trace` method show up as deprecated

Reply:
It shows deprecation in favor of `androidx.tracing`, but it's not added as a dependency to various modules.
WDYT of adding `implementation(libs.androidx.tracing.ktx)` to the convention plugins as a dependency, so we have it consistent everywhere?

Reply:
Ok, added the dependencies to android library + feature and refactored to be used everywhere. We need to use `alpha` version of the library to be able to use it with composables, but there aren't any changes in the library with the alpha (actually filed a bug to stabilize it)